### PR TITLE
*: Support the operation of `cancel DDL jobs`

### PR DESCRIPTION
**Pull request by zimulala**
If the DDL job state is done or synced, we can't cancel this job.

### CODE REVIEW

**Comment by shenli** [comment]
@coocood @winkyao PTAL

File: parser/parser.y [hunk 1]
```diff
@@ -687,7 +688,8 @@ import (
 	OptBinary		"Optional BINARY"
 	OptCharset		"Optional Character setting"
 	OptCollate		"Optional Collate setting"
-	NUM			"numbers"
+	NUM			"A number"
```
Comment (winkyao) [inline parser/parser.y]: NUM --> Number ?

File: inspectkv/inspectkv.go [hunk 2]
```diff
@@ -64,6 +64,47 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	errs := make([]error, len(ids))
+	t := meta.NewMeta(txn)
+	for i, id := range ids {
+		done := false
+		for j, job := range jobs {
+			if id != job.ID {
```
Comment (winkyao) [inline inspectkv/inspectkv.go]: just continue? add some Log here?

File: inspectkv/inspectkv.go [hunk 3]
```diff
@@ -64,6 +64,47 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
...
+				continue
+			}
+			done = true
+			// These states can't be cancelled.
+			if job.IsDone() || job.IsSynced() {
+				errs[i] = errors.New("This job will be done, so can't be cancelled")
```
Comment (winkyao) [inline inspectkv/inspectkv.go]: `This job is finished, so can't be cancelled`?

File: inspectkv/inspectkv.go [hunk 4]
```diff
@@ -64,6 +64,47 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
...
+
+	jobs, err := GetDDLJobs(txn)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	errs := make([]error, len(ids))
+	t := meta.NewMeta(txn)
+	for i, id := range ids {
+		done := false
```
Comment (winkyao) [inline inspectkv/inspectkv.go]: done --> found maybe better?

File: inspectkv/inspectkv.go [hunk 5]
```diff
@@ -64,6 +64,47 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
...
+			done = true
+			// These states can't be cancelled.
+			if job.IsDone() || job.IsSynced() {
+				errs[i] = errors.New("This job will be done, so can't be cancelled")
+				continue
+			}
+			if job.IsCancelled() {
+				continue
+			}
+			job.State = model.JobCancelling
```
Comment (coocood) [inline inspectkv/inspectkv.go]: What if job state is Rollback?

File: inspectkv/inspectkv.go [hunk 6]
```diff
@@ -64,6 +64,47 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
 	return info, nil
 }
 
+// CancelJobs cancels the DDL jobs.
+func CancelJobs(txn kv.Transaction, ids []int64) ([]error, error) {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply inspectkv/inspectkv.go]: Other functions that handle jobs are placed in this package, so I put it here.

**Comment by zimulala** [comment]
@coocood  We will rollback if the error is ErrKeyExists. It's the old logical.

File: ddl/ddl.go [hunk 7]
```diff
@@ -226,6 +227,8 @@ type ddl struct {
 	reorgDoneCh chan error
 	// reorgRowCount is for reorganization, it uses to simulate a job's row count.
 	reorgRowCount int64
+	// notifyCancelReorgJob is for reorganization, it used to notify the fill-back goroutine if the DDL job is cancelled.
+	notifyCancelReorgJob chan struct{}
```
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl.go]: Done.

File: ddl/ddl_worker_test.go [hunk 8]
```diff
@@ -245,13 +246,16 @@ func doDDLJobErr(c *C, schemaID, tableID int64, tp model.ActionType, args []inte
...
 	var checkErr error
 	if test.cancelState == job.SchemaState {
-		errs, err := inspectkv.CancelJobs(txn, test.jobIDs)
-		if err != test.cancelRetErr {
-			checkErr = errors.Trace(err)
-		}
-		// It only tests cancel one DDL job.
-		if errs[0] != test.cancelRetErrs[0] {
-			checkErr = errors.Trace(errs[0])
+		if test.act == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0 {
```
Comment (coocood) [inline ddl/ddl_worker_test.go]: This condition is too long, need to define a variable like
```
addIndexFirstReorg := test.act == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0
```
Add add comment about why ignore this state.

File: ddl/ddl_worker_test.go [hunk 9]
```diff
@@ -218,22 +221,179 @@ func testCheckJobCancelled(c *C, d *ddl, job *model.Job) {
...
+	tc.onJobUpdated = func(job *model.Job) {
+		hookCtx := mock.NewContext()
+		hookCtx.Store = store
+		var err error
+		err = hookCtx.NewTxn()
+		if err != nil {
+			checkErr = errors.Trace(err)
+		}
+		switch job.SchemaState {
+		case model.StateDeleteReorganization:
```
Comment (coocood) [inline ddl/ddl_worker_test.go]: Can we merge those cases into one line?

File: ddl/ddl_worker_test.go [hunk 10]
```diff
@@ -218,22 +221,179 @@ func testCheckJobCancelled(c *C, d *ddl, job *model.Job) {
...
+	nonexist := model.SchemaState(255)
+	err := errCancelledDDLJob
+	errs := []error{err}
+	noErrs := []error{nil}
+	tests := []testCancelJob{
+		{act: model.ActionCreateTable, jobIDs: []int64{firstID}, cancelRetErr: err, cancelRetErrs: noErrs, cancelState: model.StateDeleteOnly, nextState: model.StateWriteOnly, ddlRetErr: err},
+
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 1}, cancelRetErrs: errs, cancelState: model.StateDeleteOnly, nextState: model.StateWriteOnly, ddlRetErr: err},
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 2}, cancelRetErrs: errs, cancelState: model.StateWriteOnly, nextState: model.StateWriteReorganization, ddlRetErr: err},
+		{act: model.ActionAddIndex, jobIDs: []int64{firstID + 3}, cancelRetErrs: errs, cancelState: model.StateWriteReorganization, nextState: model.StatePublic, ddlRetErr: err},
```
Comment (coocood) [inline ddl/ddl_worker_test.go]: If the add index is converted to rollback job, the nex schema state should be StateDeleteOnly.

File: ddl/reorg.go [hunk 11]
```diff
@@ -91,10 +91,17 @@ func (d *ddl) runReorgJob(job *model.Job, f func() error) error {
 
 func (d *ddl) isReorgRunnable() error {
 	if d.isClosed() {
-		// worker is closed, can't run reorganization.
+		// Worker is closed that can't run reorganization.
```
Comment (shenli) [inline ddl/reorg.go]: Worker is closed. So it can't do reorganization job.

File: executor/builder.go [hunk 12]
```diff
@@ -143,6 +145,19 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
 	}
 }
 
+func (b *executorBuilder) buildCancelDDLJobs(v *plan.CancelDDLJobs) Executor {
+	e := &CancelDDLJobsExec{
+		baseExecutor: newBaseExecutor(v.Schema(), b.ctx),
+		JobIDs:       v.JobIDs,
+	}
+	e.errs, b.err = inspectkv.CancelJobs(e.ctx.Txn(), e.JobIDs)
```
Comment (shenli) [inline executor/builder.go]: Why not do this in CancelDDLJobsExec.Next()?

Comment (zimulala) [reply executor/builder.go]: It seems `CancelDDLJobsExec.Next()` has no `txn`.

Comment (zimulala) [reply executor/builder.go]: We can new a `txn`  and commit this `txn` in `CancelDDLJobsExec.Next()`. But this solution will commit the previous transaction.

**Comment by shenli** [comment]
@coocood @winkyao PTAL

**Comment by zimulala** [comment]
PTAL @coocood @winkyao @shenli

File: ddl/ddl_worker.go [hunk 13]
```diff
@@ -248,6 +248,18 @@ func (d *ddl) runDDLJob(t *meta.Meta, job *model.Job) (ver int64) {
 	if job.IsFinished() {
 		return
 	}
+	// The cause of this job state is that the job is cancelled by client.
+	if job.IsCancelling() {
+		// If the value of SnapshotVer isn't zero, it means the work is backfilling the indexes.
+		if job.Type == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer != 0 {
+			d.notifyCancelReorgJob <- struct{}{}
```
Comment (coocood) [inline ddl/ddl_worker.go]: How about ActionDropIndex?
And I'm not sure if there is a problem but use a `select` to send notify channel is safer.

Comment (winkyao) [reply ddl/ddl_worker.go]: how about use a map or array like:

```
m := map[int]bool{
	model.ActionAddIndex: true,
	...
}
```
to manager which job type will do reorganize. and avoid omission.

Comment (zimulala) [reply ddl/ddl_worker.go]: Now we only one action needs to do reorganize, so we don't need a map or a array.

**Comment by zimulala** [comment]
PTAL @coocood

File: ddl/index.go [hunk 14]
```diff
@@ -321,11 +320,22 @@ func (d *ddl) convert2RollbackJob(t *meta.Meta, job *model.Job, tblInfo *model.T
...
+	if err1 != nil {
+		return ver, errors.Trace(err1)
+	}
+
+	if kv.ErrKeyExists.Equal(err) {
+		return ver, kv.ErrKeyExists.Gen("Duplicate for key %s", indexInfo.Name.O)
 	}
-	return ver, kv.ErrKeyExists.Gen("Duplicate for key %s", indexInfo.Name.O)
+
+	// Clean up the channel of notifyCancelReorgJob. Make sure it can't affect other jobs.
```
Comment (coocood) [inline ddl/index.go]: It is possible that the reorg just finished, returned without any error, then the channel may affect the next job.

Move the cleanup code right after timeout error check would be safer.

**Comment by coocood** [comment]
LGTM

File: ddl/ddl.go [hunk 15]
```diff
@@ -226,6 +229,8 @@ type ddl struct {
 	reorgDoneCh chan error
 	// reorgRowCount is for reorganization, it uses to simulate a job's row count.
 	reorgRowCount int64
+	// notifyCancelReorgJob is for reorganization, it used to notify the fill-back goroutine if the DDL job is cancelled.
```
Comment (ngaut) [inline ddl/ddl.go]: s/fill-back/backfilling/

File: ddl/ddl_worker_test.go [hunk 16]
```diff
@@ -213,27 +216,183 @@ func testCheckJobDone(c *C, d *ddl, job *model.Job, isAdd bool) {
...
+
+func doDDLJobErr(c *C, schemaID, tableID int64, tp model.ActionType, args []interface{},
+	ctx context.Context, d *ddl) *model.Job {
+	return doDDLJobErrWithSchemaState(ctx, d, c, schemaID, tableID, tp, args, nil)
+}
+
+func checkCancelState(txn kv.Transaction, job *model.Job, test *testCancelJob) error {
+	var checkErr error
+	addIndexFirstReorg := test.act == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0
+	// If the action is adding index and the state is writing reorganization, it want to test the case of cancelling the job when backfilling indexes.
```
Comment (ngaut) [inline ddl/ddl_worker_test.go]: s/want/wants/

File: model/ddl.go [hunk 17]
```diff
@@ -204,6 +204,16 @@ func (job *Job) IsCancelled() bool {
 	return job.State == JobCancelled || job.State == JobRollbackDone
 }
 
+// IsRollbacking returns whether the job is rollbacking or not.
+func (job *Job) IsRollbacking() bool {
+	return job.State == JobRollback
```
Comment (ngaut) [inline model/ddl.go]: s/JobRollback/JobStateRollingback/

Comment (zimulala) [reply model/ddl.go]: I will modify the prefix in the next PR.

**Pull request merged**